UserTools/DirectionGridCheck/DirectionGridCheck.cpp

```diff
@@ -0,0 +1,245 @@
+        return true;
+    }
+
+    // Only check this event
+    //if (showEvent > 0 && (int)fEventNumber != showEvent) return true;
+    if (eventCount > 9)return true;
+
+    VertexGeometry* myvtxgeo = VertexGeometry::Instance();
+    Direction tempdir;
+
+
+    if (eventCount == 0) {
+        for (int l = 0; l < 100; l++) {
+            double phi = (6 * TMath::Pi() / 50) * l;
+            double theta = (TMath::Pi() / 200) * l;
+
+            while (phi > 2 * TMath::Pi()) {
+                std::cout << "correcting theta" << endl;
+                phi -= 2 * TMath::Pi();
+            }
+
+            tempdir.SetTheta(theta);
+            tempdir.SetPhi(phi);
+            AngularPlot->SetPoint(l, phi, theta);
+            std::cout << "Angular Plot set:" << l << ", " << phi << ", " << theta << endl;
+        }
+    }
+    RecoVertex* truevtx = 0;
+    RecoVertex* recovtx = 0;
+    auto get_muonMC = m_data->Stores.at("RecoEvent")->Get("TrueVertex", truevtx);
+    auto get_muonReco = m_data->Stores.at("RecoEvent")->Get("ExtendedVertex", recovtx);
+    double truePhi = truevtx->GetDirection().GetPhi();
+    double trueTheta = truevtx->GetDirection().GetTheta();
+    double RecoDirX = recovtx->GetDirection().X();
+    double RecoDirY = recovtx->GetDirection().Y();
+    double RecoDirZ = recovtx->GetDirection().Z();
+    double recoPhi = atan(RecoDirY/RecoDirX);
+    double recoTheta = asin(RecoDirX * RecoDirX + RecoDirY * RecoDirY);
+    if (trueTheta < 0) trueTheta += 2*TMath::Pi();
+    if (trueTheta > 2 * TMath::Pi()) trueTheta -= 2 * TMath::Pi();
+    TrueDirPlot->SetPoint(eventCount, truePhi, trueTheta);
+    std::cout << "True Plot set: " << eventCount << ", " << truePhi << ", " << trueTheta << endl;
+    Direction MRDDirection = this->findDirectionMRD();
+    MRDDirPlot->SetPoint(eventCount, MRDDirection.GetPhi(), MRDDirection.GetTheta());
+    std::cout << "MRD Plot set: " << eventCount << ", " << MRDDirection.GetPhi() << ", " << MRDDirection.GetTheta() << endl;
+    RecoDirPlot->SetPoint(eventCount, recoPhi, recoTheta);
+    std::cout << "recoPlot set: " << eventCount << ", " << recoPhi << ", " << recoTheta << endl;
+
+    eventCount++;
+  return true;
+}
+
+
+bool DirectionGridCheck::Finalise(){
+    fOutput_tfile->cd();
+    AngularPlot->Write("Angular Grid");
+    TrueDirPlot->Write("trueDirection");
+    MRDDirPlot->Write("MRDDirection");
+    RecoDirPlot->Write("recoDirection");
+    fOutput_tfile->Write();
+    fOutput_tfile->Close();
+  return true;
+}
+
+Direction DirectionGridCheck::findDirectionMRD() {
+	std::vector<BoostStore>* Tracks;
+    int numtracksinev;
+	m_data->Stores["MRDTracks"]->Get("MRDTracks", Tracks);
+	m_data->Stores["MRDTracks"]->Get("NumMrdTracks", numtracksinev);
+
+	if (numtracksinev > 1) Log("Multiple tracks need work; just using first for now", v_debug, verbosity);
+	double gradx, grady, theta, phi;
+	Direction startVertex, endVertex, result;
+	BoostStore* thisTrack = &(Tracks->at(0));
```

Comment:
What if there aren't any MRD tracks? Tracks will be an empty vector, and de-referencing thistrack will segfault.